midend/replaceSelectRange.cpp

```diff
@@ -0,0 +1,177 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+
+#include "replaceSelectRange.h"
+
+namespace P4 {
+
+static int trailingZeros(big_int n, int width) {
+    int zeros = 0;
+
+    while (n > 0 && (n & 1) == 0) {
+        zeros += 1;
+        n >>= 1;
+    }
+
+    return (zeros < width) ? zeros : width;
+}
+
+bool DoReplaceSelectRange::checkRange(const IR::Range* range) {
+    if (range == nullptr)
+        return false;
+
+    if (!range->left->is<IR::Constant>()) {
+        std::cout << "must evaluate to a compile-time "
+                  << "constant" << range->left << std::endl;
+        return false;
+    }
+    auto left = range->left->to<IR::Constant>()->value;
+    if (!range->right->is<IR::Constant>()) {
+        std::cout << "must evaluate to a compile-time "
+                  << "constant" << range->right <<  std::endl;
+        return false;
+    }
+    auto right = range->right->to<IR::Constant>()->value;
+    if (right < left) {
+        std::cout << "Range end " << std::hex << std::showbase << right
+        << " is less than start " << left << std::endl;
+        return false;
+    }
+    return true;
+}
+
+std::vector<const IR::Mask *>
+DoReplaceSelectRange::rangeToMasks(const IR::Range *r) {
+    bool st = checkRange(r);
+    BUG_CHECK(st, "Range check failed");
+
+    int width = typeMap->getType(r, true)->width_bits();
+    big_int min = r->left->to<IR::Constant>()->value;
+    big_int max = r->right->to<IR::Constant>()->value;
+    std::vector<const IR::Mask *> masks;
+    big_int range_size_remaining = max - min + 1;
+
+    while (range_size_remaining > 0) {
+        big_int range_size = ((big_int) 1) << trailingZeros(min, width);
+
+        while (range_size > range_size_remaining)
+            range_size >>= 1;
+
+        auto constType = typeMap->getType(r->left, true);
+        auto valConst = new IR::Constant(constType, min, 16, true);
+        auto mask = ~(range_size - 1) & ((((big_int) 1) << width) - 1);
+        auto maskConst = new IR::Constant(constType, mask, 16, true);
+        auto m = new IR::Mask(r->srcInfo, valConst, maskConst);
+
+        masks.push_back(m);
+        range_size_remaining -= range_size;
+        min += range_size;
+    }
+
+    return masks;
+}
+
+std::vector<IR::Vector<IR::Expression>>
+DoReplaceSelectRange::cartesianAppend(std::vector<IR::Vector<IR::Expression>> vecs,
+                                      std::vector<const IR::Mask *> masks) {
+    std::vector<IR::Vector<IR::Expression>> newVecs;
+
+    if (vecs.empty()) {
```

Comment:
is this case necessary?

Reply:
This is necessary because when the first Cartesian product is taken, the left vector is empty. Without explicitly handling this case, the resulting product would be another empty vector.